File: app/src/main/java/pl/adriankremski/coolector/network/AppDisposableObserver.java
package pl.adriankremski.coolector.network;

import com.jakewharton.retrofit2.adapter.rxjava2.HttpException;

import io.reactivex.observers.DisposableObserver;

public class AppDisposableObserver<T> extends DisposableObserver<T> {
    @Override
    public void onNext(T value) {

    }

    @Override
    public void onError(Throwable e) {
        if (e instanceof HttpException) {
            onNetworkError();
        }
    }

    public void onNetworkError() {

    }

    @Override
    public void onComplete() {

    }
}

File: app/src/main/java/pl/adriankremski/coolector/network/AppDisposableObserver.kt
package pl.adriankremski.coolector.network

import io.reactivex.observers.DisposableObserver
import pl.adriankremski.coolector.model.OperationError
import java.io.IOException
import java.net.UnknownHostException

open class AppDisposableObserver<T> : DisposableObserver<T>() {
    override fun onNext(value: T) {

    }

    override fun onError(e: Throwable) {
        if (e is OperationError) {
            onServerError(e.operation.message)
        } else if (e is UnknownHostException || e is IOException) {
            onNetworkError()
        }
    }

    open fun onServerError(message: String?) {

    }

    open fun onNetworkError() {

    }

    override fun onComplete() {

    }
}
